### DISCUSSION TITLE
How to remotely mount Rook-Ceph persistent volumes outside of Kubernetes using the ceph-fuse client or the mount -t client

### DISCUSSION THREAD

**Discussion by wangchao6856** [rook/rook#14501] [Q&A] [answered]
A Rook cluster has been successfully deployed on Kubernetes, and all functionalities are working correctly. 

Now, we need to investigate how to use the Ceph client to mount volumes outside of Kubernetes. 

The current test results are as follows: 
A pod was started in a different namespace within the Kubernetes cluster, and within the pod, the ceph-fuse client was downloaded and the /etc/ceph/ceph.conf information was configured, with mon_host set to <rook-ceph-mon internal IP>. 
Using the ceph-fuse command, the mount was successful. 
However, using the external export IP and NodePort of the rook-ceph-mon service resulted in an error. 
The routing policy for the exported service is externalTrafficPolicy: Cluster, with the port type as NodePort. 
The error message is as follows:
```
2024-07-26T03:46:57.922+0000 7fb06d8e0080 -1 init, newargv = 0x55b3872dbae0 newargc=9
ceph-fuse[86]: starting ceph client
2024-07-26T03:46:57.926+0000 7fb0667fc700 -1 monclient(hunting): handle_auth_bad_method server allowed_methods [2] but i only support [2]
2024-07-26T03:46:57.926+0000 7fb0657fa700 -1 monclient(hunting): handle_auth_bad_method server allowed_methods [2] but i only support [2]
2024-07-26T03:46:57.926+0000 7fb065ffb700 -1 monclient(hunting): handle_auth_bad_method server allowed_methods [2] but i only support [2]
ceph-fuse[86]: ceph mount failed with (13) Permission denied
2024-07-26T03:46:57.930+0000 7fb06d8e0080 -1 client.0 authentication failed: (13) Permission denied
2024-07-26T03:46:57.930+0000 7fb06d8e0080 -1 client.0 mdsmap subscription failed: (13) Permission denied
```
the ceph-fuse command with mon_host external export ip is:

`ceph-fuse /mnt/mycephfs -m [v2:<export-ip-mon-c:nodeport>,v1:<export-ip-mon-c:nodeport>],[v2:<export-ip-mon-a:nodeport>,v1:<export-ip-mon-a:nodeport>],[v2:<export-ip-mon-b:nodeport>,v1:<export-ip-mon-b:nodeport>] -n client.admin -k /etc/ceph/client.admin.keyring --client_fs=ceph-filesystem  --no-mon-config`

Regardless of whether the --no-mon-config parameter is used, the mount fails, and the admin keyring is correct.

If the -m parameter is not specified for ceph-fuse, and mon_host is configured as an external address in /etc/ceph/ceph.conf, the following error occurs:
`monclient: get_monmap_and_config failed to get config`

Can someone provide assistance? Thank you.

**Comment by Madhu-1** [comment]
https://rook.io/docs/rook/latest-release/Troubleshooting/direct-tools/?h=direct#shared-filesystem-tools should give you an idea for it.

**Reply by wangchao6856** [reply]
I understand that this article discusses using the mount -t command to link to the internal IP of rook-ceph in Kubernetes. However, my goal is to figure out how to mount rook-ceph's persistent volume outside of Kubernetes

**Reply by Madhu-1** [reply]
the idea to access the mount will remain the same, all depends on which networking you run the ceph cluster, below i provided link to run ceph on host networking as well

**Comment by wangchao6856** [comment]
Has anyone solved this problem?

**Reply by Madhu-1** [reply]
@wangchao6856 you can run ceph cluster with HostNetworking if you want to use volumes outside of Kubernetes cluster. https://rook.io/docs/rook/latest-release/CRDs/Cluster/network-providers/#host-networking

**Reply by wangchao6856** [reply]
I have tested this solution before by enabling this setting in the rook-cluster YAML, but it was still ineffective. I plan to try again using this solution

**Reply by wangchao6856** [reply]
I redeployed Rook and set network.provider: host, but according to the Helm chart source code, this parameter is used in the Deployment of rook-ceph-tools and is not related to the mon service itself. Additionally, after redeploying, this setting did not work for accessing outside of Kubernetes

**Reply by wangchao6856** [reply]
I also tested setting the mon pods to hostNetwork, but this caused the osd to become unusable because the mon no longer had an internal IP

**Reply by travisn** [reply]
> I also tested setting the mon pods to hostNetwork, but this caused the osd to become unusable because the mon no longer had an internal IP

Did you enable the host networking in a new Rook cluster, or did you try to convert the existing cluster? In a new cluster, the OSDs should be perfectly stable connecting to the mons with the host ip.

**Reply by wangchao6856** [reply]
Yes, I reinstalled the Rook cluster, and during the reinstallation, I set network.provider: host. How can I confirm that the communication between OSD and MON is using the host IP? Anyway, I tested the host's ports 3300 and 6789, and they are not accessible

**Reply by wangchao6856** [reply]
What I want to ask is whether Rook itself supports access from outside K8s, or if anyone has experience connecting to Rook from outside K8s. If it is supported or if someone has used it this way, then there might be some issues with my configuration.

**Reply by wangchao6856** [reply]
Another experiment is that we have a Ceph deployment directly on a physical machine, and it has been tested to support remote client mounting. The reason I can think of is that the physical machine deployment does not have the issue of internal and external IP addresses.

**Reply by zhucan** [reply]
`kubectl get pods -o wide -n rook-ceph | grep mon` can you pasted  the results?  and check the pods whether use the hostnetwork, then check the iptables.

**Reply by wangchao6856** [reply]
![Dingtalk_20240731154818](https://github.com/user-attachments/assets/6466cf0c-342f-48bb-9429-d9c46b3dc721)

**Reply by zhucan** [reply]
ip of your node is 10.233.x.x?

**Reply by wangchao6856** [reply]
this is pod ip，not node ip

**Reply by zhucan** [reply]
so you didn't use hostnetwork  mode,  so you can not access the mon of ceph  out of the ceph cluster（k8s cluster）.

**Reply by zhucan** [reply]
![image](https://github.com/user-attachments/assets/37cf121c-39ef-44e5-856f-5cb25d74f18b)
like this.

**Reply by wangchao6856** [reply, marked as answer]
Do you mean to specify hostnetwork in the network of ceph-cluster's values.yaml? I can try it。I think the official documentation link for this parameter is:
https://rook.github.io/docs/rook/latest-release/CRDs/specification/?h=hostnetwork#ceph.rook.io/v1.NetworkSpec

**Reply by zhucan** [reply]
yes

**Comment by wangchao6856** [comment]
Thank you all for your help. After testing, by setting hostNetwork in values.yaml, the mon, osd, and mgr modules will start using the host network mode. The mon_host will use the host machine's IP, thus avoiding the issue of having different internal and external IPs. Testing has confirmed that it is successfully to mount outside of K8s using the ceph-fuse client and the mount -t method.

**Reply by ToolKami** [reply]
Hi @wangchao6856 , I am facing a similar issue here at https://github.com/rook/rook/discussions/15611, is setting hostNetwork in `values.yaml` versus `cluster.yaml` different? Can you kindly share your setup? Thanks!
